Sparse index: `checkout-index`

vdye opened the pull request:
## Changes
- Expand testing in `t1092` to cover `checkout-index`
- Modify the behavior of `checkout-index --all` to make more sense with sparse checkouts (sparse index or not)
  - Introduce `--sparse` flag maintaining "old" behavior
- Integrate sparse index with `checkout-index`, including corresponding `ensure_not_expanded` tests

vdye committed:
checkout-index: expand sparse checkout compatibility tests

Add tests to cover basic `checkout-index` cases related to sparse checkouts:
files and folder, both inside and outside of sparse checkout definition. New
tests will serve as a baseline for expected behavior when integrating
`checkout-index` with the sparse index.

Of note is the test demonstrating the behavior of `checkout-index --all`;
creating files even outside the sparse checkout definition is somewhat
unintuitive, and will cause significant performance issues when run with a
sparse index. The goal of later changes will be to change this default
behavior and introduce a `--sparse` flag to enable populating files outside
the sparse definition.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
checkout-index: add `--sparse` option

Change the default behavior of `checkout-index --all` for sparse checkouts
to no longer refresh files outside the sparse checkout definition. The
newly-added `--sparse` option, when used with `--all`, maintains the "old"
behavior and checks out files outside the sparse checkout definition.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
checkout-index: integrate with sparse index

Add repository settings to allow usage of the sparse index. In order to prevent
unexpected errors when attempting to check out a sparse directory entry,
`checkout_file` directly checks whether a found entry is a sparse directory
and, if so, exits with an error. The test corresponding to this case now
verifies the error message, intentionally differing from the non-sparse index
scenarios.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
checkout-index: improve `ensure_full_index` scope in `checkout_all`

Update `checkout_all` to only run `ensure_full_index` when entries in
sparse directories are needed (i.e., `--sparse` is specified).

Signed-off-by: Victoria Dye <vdye@github.com>

ldennington commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -700,6 +700,60 @@ test_expect_success 'cherry-pick with conflicts' '
+	init_repos &&
+
+	run_on_all rm -f deep/a &&
+	test_all_match git checkout-index -- deep/a &&
```
I've seen this `--` usage several times but am not quite sure what it does. Mind providing a quick explanation?

derrickstolee replied:
The `--` signals the end of options and everything that remains should be parsed as plain-text input for a list of values. Typically, this means a list of pathspecs. It allows you to do something like `git checkout-index -- --looks-like-an-option` if you are silly enough to have `--looks-like-an-option` in your working tree.

ldennington commented on Documentation/git-checkout-index.txt:
```diff
@@ -59,6 +61,10 @@ OPTIONS
 	write the content to temporary files.  The temporary name
 	associations will be written to stdout.
 
+--sparse::
```
This feels a little odd. From what I understand, we generally use the `--sparse` flag to enable sparse behavior, but here we seem to be circumventing it. We may want to consider something more like `--ignore-sparse`. Let me know if I'm missing something, though!

derrickstolee replied:
Check out gitgitgadget/git#1018 which restricts how Git operates outside of the sparse-checkout. The `--sparse` option allows overriding these protections. It's _really new_.

vdye replied:
@ldennington I definitely see where you're coming from (it does seem somewhat backwards that `--sparse` would expand the index/do things outside the sparse checkout). I did base it on @derrickstolee's linked PR (+`ls-trees`), and I think this line conveys it best:

> a new '--sparse' option is added that ignores the sparse-checkout patterns and the SKIP_WORKTREE bit

The option here is pretty much treating `--sparse` as "does things in parts of the repository that should be sparse", rather than ignoring those parts of the repository. @derrickstolee please correct me if I'm using it wrong, though - I'm mostly looking for this to be consistent with the other already-implemented options (and offer a way to use `checkout-index --all` without always expanding the full index).

derrickstolee replied:
@vdye: I think you have the right approach. We should make Git ignore files outside of the sparse-checkout cone as much as possible, and allow users to override that behavior by specifying `--sparse`. We do not expect users to have a legitimate reason to care about files outside of the sparse-checkout, but Git needs to care unless we make these changes.

ldennington commented on Documentation/git-checkout-index.txt:
```diff
@@ -37,8 +38,9 @@ OPTIONS
-	checks out all files in the index.  Cannot be used
-	together with explicit filenames.
+	checks out all files in the index, excluding those outside
+	any specified sparse checkout patterns (see `--sparse`).
```
Reminder to update this if you decide to take below suggestion.

ldennington commented on Documentation/git-checkout-index.txt:
```diff
@@ -12,6 +12,7 @@ SYNOPSIS
 'git checkout-index' [-u] [-q] [-a] [-f] [-n] [--prefix=<string>]
 		   [--stage=<number>|all]
 		   [--temp]
+		   [--sparse]
```
Reminder to update this if you decide to take below suggestion.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -700,6 +700,60 @@ test_expect_success 'cherry-pick with conflicts' '
+
+	test_all_match git checkout-index --all &&
+	test_sparse_match test_path_is_missing folder1
+'
```
Good catch! Scary that this could completely abolish all performance benefits for a sparse-checkout!

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -741,7 +741,14 @@ test_expect_success 'checkout-index with folders' '
 
 	# Outside checkout definition
-	test_all_match test_must_fail git checkout-index -f -- folder1/
+	# Note: although all tests fail (as expected), the messaging differs. For
+	# non-sparse index checkouts, the error is that the "file" does not appear
+	# in the index; for sparse checkouts, the error is explicitly that the
+	# entry is a sparse directory.
```
Is it possible to make both implementations say "file is outside the sparse-checkout definition"?

* The "file does not appear in the index" is not entirely correct for full indexes.
* The "file is a sparse directory" has a bit too much internal information.

I wonder if using `if (!path_in_sparse_checkout(name, &the_index))` somewhere would help here.

It's perhaps a bit tricky how this loop is structured, since it is doing a prefix match on a range of index entries.

In this test script, what happens with this example?

```
git sparse-checkout set deep/deeper1
git checkout-index -f -- deep/
```

Note that `deep/deeper2/` is a sparse directory within the `deep/` directory, but we would still want the other entries to succeed, silently ignoring the `deep/deeper2/` entry. This would only be an _error_ if the user ran `git checkout-index -f -- deep/deeper2/`, right?

vdye replied:
I think it's just comparing the names, not performing a prefix match (the first condition in this check makes sure the names are the same length):

```
if (ce_namelen(ce) != namelen ||
		    memcmp(ce->name, name, namelen))
```

The result is that, from your example, `git checkout-index -f -- deep/` (or any path that doesn't refer to a file entry in the index) returns `git checkout-index: <name> is not in the cache`. It seems like this command expects _exact_ names of cache entries (except for the subcommands using a pathspec, but they handle things differently anyway), in which case `<name> is not in the cache` makes some amount of sense? Agreed that it's misleading, though - would that be something that should be clarified in the command doc, or in this message itself?

As for checking whether something is in the sparse checkout, this doesn't actually block checking out files outside the sparse definition (you can successfully run `git checkout-index -- folder1/a` without any other repo setup). The issue is that sparse directories _do_ appear as standalone cache entries, so without the `S_ISSPARSEDIR` check the entry would be passed along to `checkout_entry` (which then fails because it doesn't handle directories). Since there's no support `checkout-index`-ing directories otherwise, this probably needs to be blocked rather than worked around. Maybe the message should be the same as whenever anyone tries to `checkout-index` a directory in general? I.e., "lie" and say `<name> is not in the cache`, or some other error message that's the same between them?

derrickstolee commented on builtin/checkout-index.c:
```diff
@@ -247,6 +253,8 @@ int cmd_checkout_index(int argc, const char **argv, const char *prefix)
 
 		if (all)
 			die("git checkout-index: don't mix '--all' and explicit filenames");
+		if (include_sparse)
+			die("git checkout-index: don't mix '--sparse' and explicit filenames");
```
Why shouldn't these be combined? What if I want to poke at a file outside of my sparse-checkout? It certainly seems like `git checkout-index --sparse -- deep/deeper2/a` would be a good way to get that file on-disk.

vdye replied:
The reason I didn't include `--sparse` for individual files is because, as it is right now, individually-specified files will always expand the index if that specified file is not already in the index - I wasn't really sure what the "right" way to handle it was, though. On one hand, not requiring a `--sparse` flag to `checkout-index` individual files (even those outside the checkout definition) is the most consistent with existing behavior. On the other hand, it is kind of weird that we need a special `--sparse` option for files in sparse directories for `--all`, but not when you list out individual files.

What do you think would be better (especially based on how you've handled similar situations in other commands)?

derrickstolee replied:
Since this is a plumbing command, let's be cautious and avoid the behavior break by requiring `--sparse` for individual files. (This is a change of opinion from my earlier comment.)

derrickstolee reviewed (commented):
Good progress. I have a few questions.

derrickstolee reviewed (approved):
All of my prior concerns were misconceptions about how `git checkout-index` works. Thanks for clearing them up!

vdye committed:
checkout-index: add performance test for `checkout-index --all`

`git checkout-index --all` is a subcommand of `git stash`, so it is helpful
to verify the performance of that particular usage.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye commented:
@derrickstolee + @ldennington I added one last change (performance test for `git checkout-index -f --all`), which had these results (tip of `vfs-2.33.0` vs the tip of this branch):

```
Test                                                   ms/vfs-2.33.0       HEAD                  
-------------------------------------------------------------------------------------------------
2000.30: git checkout-index -f --all (full-v3)         76.79(9.89+64.59)   0.52(0.22+0.26) -99.3%
2000.31: git checkout-index -f --all (full-v4)         80.96(9.65+63.89)   0.52(0.21+0.26) -99.4%
2000.32: git checkout-index -f --all (sparse-v3)       89.43(9.83+64.99)   0.34(0.05+0.24) -99.6%
2000.33: git checkout-index -f --all (sparse-v4)       93.67(9.65+63.46)   0.34(0.05+0.25) -99.6%
```

The huge speedup seems to mostly come from having `checkout-index --all` respect the sparse checkout definition, but there is also improvement from full index to sparse index (which is nice to see)!

derrickstolee commented:
> @derrickstolee + @ldennington I added one last change (performance test for `git checkout-index -f --all`), which had these results (tip of `vfs-2.33.0` vs the tip of this branch):
> 
> ```
> Test                                                   ms/vfs-2.33.0       HEAD                  
> -------------------------------------------------------------------------------------------------
> 2000.30: git checkout-index -f --all (full-v3)         76.79(9.89+64.59)   0.52(0.22+0.26) -99.3%
> 2000.31: git checkout-index -f --all (full-v4)         80.96(9.65+63.89)   0.52(0.21+0.26) -99.4%
> 2000.32: git checkout-index -f --all (sparse-v3)       89.43(9.83+64.99)   0.34(0.05+0.24) -99.6%
> 2000.33: git checkout-index -f --all (sparse-v4)       93.67(9.65+63.46)   0.34(0.05+0.25) -99.6%
> ```
> 
> The huge speedup seems to mostly come from having `checkout-index --all` respect the sparse checkout definition, but there is also improvement from full index to sparse index (which is nice to see)!

Yeah!  Those 75-90s times are definitely from writing too much data to the working tree. It's nice to see exactly why this is such an important change to include and why users would not like that old behavior!
